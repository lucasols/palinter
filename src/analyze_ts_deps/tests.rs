use std::collections::BTreeMap;

use super::*;
use insta::assert_debug_snapshot;

struct SimplifiedFile {
    path: PathBuf,
    content: String,
}

fn create_flatten_root_structure(
    files: Vec<SimplifiedFile>,
) -> HashMap<String, File> {
    let mut flatten_root_structure = HashMap::new();

    for file in files {
        let file_to_add = File {
            basename: file.path.file_stem().unwrap().to_str().unwrap().to_string(),
            name_with_ext: file
                .path
                .file_name()
                .unwrap()
                .to_str()
                .unwrap()
                .to_string(),
            content: Some(file.content),
            extension: None,
            path: file.path.to_str().unwrap().to_string(),
        };

        flatten_root_structure
            .insert(file.path.to_str().unwrap().to_string(), file_to_add);
    }

    flatten_root_structure
}

#[test]
fn get_project_files_deps_info_test() {
    let entry_points = vec![PathBuf::from("@src/index.ts")];

    let flattened_root_structure = create_flatten_root_structure(vec![
        SimplifiedFile {
            path: PathBuf::from("./src/index.ts"),
            content: String::from("import '@src/fileB';"),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileB.ts"),
            content: String::from(
                r#"
                    import { a } from '@src/fileA';
                    import { b } from '@src/fileA';
                    import { test } from 'testLib';
                    "#,
            ),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileA.ts"),
            content: String::from(
                r#"
                    export const a = 1;
                    export const b = 2;
                    "#,
            ),
        },
    ]);

    let result = get_used_project_files_deps_info(
        entry_points,
        flattened_root_structure,
        &HashMap::from_iter(vec![(String::from("@src"), String::from("./src"))]),
        &mut TsProjectCtx::default(),
    )
    .unwrap();

    assert_debug_snapshot!(BTreeMap::from_iter(result));
}

#[test]
fn get_project_files_deps_info_test_2() {
    let entry_points = vec![PathBuf::from("@src/index.ts")];

    let flattened_root_structure = create_flatten_root_structure(vec![
        SimplifiedFile {
            path: PathBuf::from("./src/index.ts"),
            content: String::from(
                r#"
                    import { a } from '@src/fileA';
                    import { b } from '@src/fileB';
                    "#,
            ),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileB.ts"),
            content: String::from(
                r#"
                    import { c } from '@src/fileC';
                    "#,
            ),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileA.ts"),
            content: String::from(
                r#"
                    import { c } from '@src/fileC';
                    "#,
            ),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileC.ts"),
            content: String::from(
                r#"
                    export const c = 1;
                    "#,
            ),
        },
    ]);

    let result = get_used_project_files_deps_info(
        entry_points,
        flattened_root_structure,
        &HashMap::from_iter(vec![(String::from("@src"), String::from("./src"))]),
        &mut TsProjectCtx::default(),
    )
    .unwrap();

    assert_debug_snapshot!(BTreeMap::from_iter(result));
}

#[test]
fn project_with_circular_deps() {
    let entry_points = vec![PathBuf::from("@src/index.ts")];

    let flattened_root_structure = create_flatten_root_structure(vec![
        SimplifiedFile {
            path: PathBuf::from("./src/index.ts"),
            content: String::from("import '@src/fileB';"),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileB.ts"),
            content: String::from(
                r#"
                    import { a } from '@src/fileA';
                    import { b } from '@src/fileA';
                    import { test } from 'testLib';
                    export const c = 2;
                    "#,
            ),
        },
        SimplifiedFile {
            path: PathBuf::from("./src/fileA.ts"),
            content: String::from(
                r#"
                    import { c } from '@src/fileB';
                    export const a = 1;
                    export const b = 2;
                    "#,
            ),
        },
    ]);

    let result = get_used_project_files_deps_info(
        entry_points,
        flattened_root_structure,
        &HashMap::from_iter(vec![(String::from("@src"), String::from("./src"))]),
        &mut TsProjectCtx::default(),
    )
    .unwrap();

    assert_debug_snapshot!(BTreeMap::from_iter(result));
}
