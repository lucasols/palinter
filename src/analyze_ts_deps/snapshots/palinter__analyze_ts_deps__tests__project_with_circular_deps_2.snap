---
source: src/analyze_ts_deps/tests.rs
expression: "BTreeMap::from_iter(result)"
---
{
    "./src/1.ts": FileDepsInfo {
        deps: {
            "./src/2.ts",
            "./src/3.ts",
            "./src/1.ts",
        },
        circular_deps: Some(
            [
                "|./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
            ],
        ),
        imports: {
            "./src/2.ts": Import {
                import_path: "@src/2.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/2.ts": FileDepsInfo {
        deps: {
            "./src/3.ts",
            "./src/1.ts",
            "./src/2.ts",
        },
        circular_deps: Some(
            [
                "|./src/2.ts| > ./src/3.ts > ./src/1.ts > |./src/2.ts|",
            ],
        ),
        imports: {
            "./src/3.ts": Import {
                import_path: "@src/3.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/3.ts": FileDepsInfo {
        deps: {
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
        },
        circular_deps: Some(
            [
                "|./src/3.ts| > ./src/1.ts > ./src/2.ts > |./src/3.ts|",
            ],
        ),
        imports: {
            "./src/1.ts": Import {
                import_path: "@src/1.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/b.ts": FileDepsInfo {
        deps: {
            "./src/d.ts",
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
            "./src/e.ts",
            "./src/f.ts",
            "./src/b.ts",
        },
        circular_deps: Some(
            [
                "./src/b.ts > ./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
                "|./src/b.ts| > ./src/e.ts > ./src/f.ts > |./src/b.ts|",
            ],
        ),
        imports: {
            "./src/d.ts": Import {
                import_path: "@src/d.ts",
                line: 1,
                values: SideEffect,
            },
            "./src/e.ts": Import {
                import_path: "@src/e.ts",
                line: 2,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/c.ts": FileDepsInfo {
        deps: {
            "./src/d.ts",
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
        },
        circular_deps: Some(
            [
                "./src/c.ts > ./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
            ],
        ),
        imports: {
            "./src/d.ts": Import {
                import_path: "@src/d.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/d.ts": FileDepsInfo {
        deps: {
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
        },
        circular_deps: Some(
            [
                "./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
            ],
        ),
        imports: {
            "./src/1.ts": Import {
                import_path: "@src/1.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/e.ts": FileDepsInfo {
        deps: {
            "./src/f.ts",
            "./src/b.ts",
            "./src/d.ts",
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
            "./src/e.ts",
        },
        circular_deps: Some(
            [
                "./src/e.ts > ./src/f.ts > ./src/b.ts > ./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
                "|./src/e.ts| > ./src/f.ts > ./src/b.ts > |./src/e.ts|",
            ],
        ),
        imports: {
            "./src/f.ts": Import {
                import_path: "@src/f.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/f.ts": FileDepsInfo {
        deps: {
            "./src/b.ts",
            "./src/d.ts",
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
            "./src/e.ts",
            "./src/f.ts",
        },
        circular_deps: Some(
            [
                "./src/f.ts > ./src/b.ts > ./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
                "|./src/f.ts| > ./src/b.ts > ./src/e.ts > |./src/f.ts|",
            ],
        ),
        imports: {
            "./src/b.ts": Import {
                import_path: "@src/b.ts",
                line: 1,
                values: SideEffect,
            },
        },
        exports: [],
    },
    "./src/index.ts": FileDepsInfo {
        deps: {
            "./src/b.ts",
            "./src/d.ts",
            "./src/1.ts",
            "./src/2.ts",
            "./src/3.ts",
            "./src/e.ts",
            "./src/f.ts",
            "./src/c.ts",
        },
        circular_deps: Some(
            [
                "./src/index.ts > ./src/b.ts > ./src/d.ts > |./src/1.ts| > ./src/2.ts > ./src/3.ts > |./src/1.ts|",
                "./src/index.ts > |./src/b.ts| > ./src/e.ts > ./src/f.ts > |./src/b.ts|",
            ],
        ),
        imports: {
            "./src/b.ts": Import {
                import_path: "@src/b.ts",
                line: 1,
                values: SideEffect,
            },
            "./src/c.ts": Import {
                import_path: "@src/c.ts",
                line: 2,
                values: SideEffect,
            },
        },
        exports: [],
    },
}
